controller/src/main/java/io/pravega/controller/server/bucket/BucketService.java

```diff
@@ -0,0 +1,276 @@
+    private final BucketStore.ServiceType serviceType;
+    private final Semaphore semaphore;
+    private final Object lock = new Object();
+    @GuardedBy("lock")
+    private final PriorityQueue<QueueElement> workQueue;
+    @GuardedBy("lock")
+    private final Set<Stream> knownStreams;
+    private final LinkedBlockingQueue<StreamNotification> notifications;
+    private final CompletableFuture<Void> serviceStartFuture;
+    private final CompletableFuture<Void> notificationLoop;
+    private final CompletableFuture<Void> workerLoop;
+    private final Duration executionPeriod;
+    private final BucketWork bucketWork;
+    private final Thread notification;
+    private final Thread worker;
+    
+    BucketService(BucketStore.ServiceType serviceType, int bucketId, ScheduledExecutorService executor,
+                  int maxConcurrentExecutions, Duration executionPeriod, BucketWork bucketWork) {
+        this.serviceType = serviceType;
+        this.bucketId = bucketId;
+        this.executor = executor;
+        this.notifications = new LinkedBlockingQueue<>();
+        this.serviceStartFuture = new CompletableFuture<>();
+        this.notificationLoop = new CompletableFuture<>();
+        this.workerLoop = new CompletableFuture<>();
+        semaphore = new Semaphore(maxConcurrentExecutions);
+        this.knownStreams = new HashSet<>();
+        this.workQueue = new PriorityQueue<>(Comparator.comparingLong(x -> x.nextExecutionTimeInMillis));
+        this.executionPeriod = executionPeriod;
+        this.bucketWork = bucketWork;
+        String threadRoot = String.format("%s-%d-", serviceType.getName(), bucketId);
+        String notificationThreadName = String.format("%s-notification", threadRoot);
+        String workerThreadName = String.format("%s-worker", threadRoot);
+        this.notification = new Thread(this::notificationLoop, notificationThreadName);
+        this.worker = new Thread(this::workerLoop, workerThreadName);
+    }
+
+    @Override
+    public void doStart() {
+        CompletableFuture.runAsync(() -> {
+            try {
+                startBucketChangeListener();
+
+                notifyStarted();
+
+                notification.start();
+                worker.start();
+            } finally {
+                log.info("{}: bucket {} service start completed", getServiceType(), getBucketId());
+                serviceStartFuture.complete(null);
+            }
+        });
+    }
+
+    abstract void startBucketChangeListener();
+
+    abstract void stopBucketChangeListener();
+
+    private void notificationLoop() {
+        log.info("{}: Notification loop terminated for bucket {}", serviceType, bucketId);
+
+        try {
+            long executionDuration = executionPeriod.toMillis();
+            while (isRunning()) {
+                StreamNotification notification =
+                        Exceptions.handleInterruptedCall(() -> notifications.poll(100, TimeUnit.MILLISECONDS));
```

Comment:
I followed the suggestion partially. Since the intent is to not have dedicated threads that may block most of the time, i made it all asynchronous, while still using `AbstractService` so that i can manage the lifecycle of my fields in the class without having to let the base class manage the thread for me which would have been the case with `AbstractThreadPoolService `. 

I really liked the class `BlockingDrainingQueue` and have used it for notifications processing asynchronously. So an asynchronous notification loop (Futures.loop) keeps calling `take` on this queue and whenever this take completes, some work is performed with respect to the stream loop continues with the next iteration. 
This ensures notification loop is does not require a dedicated thread that sleeps.

The priority queue used in `worker loop` had different requirements - we need to submit at the most maxConcurrentExecutions into the executor's queue. So for that purpose, we maintain a separate counter instead of a semaphore as was used earlier. So there is no blocking. 
The worker loop (`Futures.loop`), in each iteration checks if there are slots available or not (counter >0). If slots are available, it `peeks` into priority queue to see if there is a stream available in the priority queue or not.